Supporting to distinguish multiple consent forms during onboarding via an identifier (#50)

RealLast opened the pull request:
# *Support for storing multiple consent forms (#50)*

## :recycle: Current situation & Problem
Closes #50 
In the case of multiple `OnboardingConsentView`, it was impossible to distinguish the individual consent forms when storing in the `Standard`, as the "store" function of the `OnboardingConstraint` only receives a `PDFDocument` as a parameter and no additional information.

## :gear: Release Notes 
* Added a parameter "identifier" to `OnboardingConsentView` and `OnboardingConstraint`. 
* If an identifier is specified when creating an `OnboardingConsentView', then the store function will be called with that identifier; otherwise the identifier is set to the default value "DefaultConsentDocument".
* UITests/TestApp has been changed accordingly to test the new functionality of having multiple (in this case two) consent forms.
    * Added "First" and "Second" Test View and Rendering View 
    * `ExampleStandard` has been changed to distinguish the two consent forms based on their identifiers "FirstConsentDocument" and "SecondConsentDocument"


## :books: Documentation
*Documented changes in the appropriate Markdown files.


## :white_check_mark: Testing
* Adapted UI Test app to include a second `OnboardingConsentView` with an identifier different from the first one.

If you accept this PR, there are implications for [SpeziTemplateApplication](https://github.com/StanfordSpezi/SpeziTemplateApplication) and [SpeziStudyApplication](https://github.com/StanfordSpezi/SpeziStudyApplication), as both use an `OnboardingConsentView` and require to add the "identifier" parameter to the respective `Standard`. I am happy to make those changes if this PR gets accepted.


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on Tests/UITests/TestApp/ExampleStandard.swift:
```diff
@@ -14,19 +14,33 @@ import SwiftUI
+    func store(consent: PDFDocument, identifier: String) async {
         await MainActor.run {
-            self.consentData = consent
+            if identifier == "FirstConsentDocument" {
```
I have thought a bit about the usage of "magic strings" as identifiers for the project.

I think it still makes sense to use strings as an extensible way to define identifiers and make it possible to load identifiers from a remote system or make them dynamic compared to having a typed solution with, e.g., enum cases.

I would nevertheless suggest that we demonstrate some best practices in the documentation and testing application. One would be to avoid re-declaring the string across the application and rather use an `enum` with no cases to represent the choices and use that across the application where we reference the identifier.

```swift
enum ConsentDocumentIdentifier {
    static let first = "consentDocumentIdentifier1"
    static let second = "consentDocumentIdentifier2"
}
```

This would then turn this statement to:
```suggestion
            if identifier == ConsentDocumentIdentifier.first {
```

RealLast replied:
Done! I added an enum ConsentDocumentIdentifier with identifier strings, instead of using the actual "magic strings" throughout the code.

PSchmiedmayer commented on Sources/SpeziOnboarding/SpeziOnboarding.docc/ObtainingUserConsent.md:
```diff
@@ -28,6 +28,7 @@ OnboardingConsentView(
     action: {
         // Action to perform once the user has given their consent
     },
+    identifier: "MyFirstConsentForm", // Specify an optional unique identifier for the consent form, helpful for distinguishing consent forms when storing.
```
It would be great if we can also use the code example as a way to explain the usage of enums to avoid "magic strings" as identifiers as noted in the other code comment.

RealLast replied:
I added an example to the document.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingDataSource.swift:
```diff
@@ -43,9 +43,9 @@ public class OnboardingDataSource: Module, EnvironmentAccessible {
     /// - Parameter consent: The exported consent form represented as `PDFDocument` that should be added.
-    public func store(_ consent: PDFDocument) async {
+    public func store(_ consent: PDFDocument, identifier: String) async {
         Task { @MainActor in
```
I am not quite sure why we dispatch to an other task here; I think we should aim to not create too many tasks between different calls and would therefore suggest to remove this here.

RealLast replied:
I agree, a separate task should not be necessary. I have removed it.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConstraint.swift:
```diff
@@ -15,6 +15,8 @@ import Spezi
+    /// - Parameters:
+    ///     - consent: The exported consent form represented as `PDFDocument` that should be added.
+    ///     - identifier: A 'String' identifying the consent form as specified in OnboardingConsentView.                
+    func store(consent: PDFDocument, identifier: String) async
```
If we change the signature here, I would suggest to make this function throwing and therefore use this as a chance to surface errors to the user and make sure that the user can't progress with the onboarding if the uploading didn't work as expected.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingDataSource.swift:
I would suggest that we make sure that this PR is a non-breaking change.

I think the easier way to introduce a `ConsentConstraint` that 
1. Has a better name to focus on the consent document
2. Also has the option to throw an error back to the consent view to do some error handling there
3. Makes it easier to do a backward-compatible implementation.
```swift
public protocol ConsentConstraint: Standard {
    func store(consent: PDFDocument, identifier: String) async throws
}
```

We can then deprecate the old constrain to surface the warning in the apps/packages using this functionality.
```swift
@available(
    *,
    deprecated,
    message: """
    Storing consent documents without an identifier is deprecated.
    Please use `ConsentConstraint` instead.
    """
)
public protocol OnboardingConstraint: Standard {
    @available(
        *,
        deprecated,
        message: """
        Storing consent documents without an identifier is deprecated.
        Please use ConsentConstraint.store(consent: PDFDocument, identifier: String) instead.
        """
    )
    func store(consent: PDFDocument) async
}
```

I would suggest to improve this data source to still support the old mechanism (at least until we tag a 2.0 or move this to a dedicated Spezi Consent Swift Package) while gently pointing people to the updated version.

I would therefore suggest to use a dynamic type checking for the `Standard` instead of using the `any` constraint in the `@StandardActor` dependency injection.

I would suggest to do something similar to this:
1. Runtime check on configure if we have a standard that conforms to one of the possible protocols
4. Runtime check in the store function to still dispatch to the old function if the `ConsentConstraint` is not yet implemented. If neither are implemented we should fail again (could never really happen due to the configure setup) but would still be good to do a double check here.

Something like this would be the right direction:
```swift
public class OnboardingDataSource: Module, EnvironmentAccessible {
    @StandardActor private var standard: any Standard
    
    
    public init() { }
    
    
    public func configure() {
        guard standard is any OnboardingConstraint || standard is any ConsentConstraint else {
            fatalError("A \(type(of: standard).self) must conform to `ConsentConstraint` to process signed consent documents.")
        }
    }
    
    public func store(_ consent: PDFDocument, identifier: String) async throws {
        if let consentConstraint = standard as? any ConsentConstraint {
            try await consentConstraint.store(consent: consent, identifier: identifier)
        } else if let onboardingConstraint = standard as? any OnboardingConstraint {
            await onboardingConstraint.store(consent: consent)
        } else {
            fatalError("A \(type(of: standard).self) must conform to `ConsentConstraint` to process signed consent documents.")
        }
    }
}
```

RealLast replied:
I agree, it would be good to have some backward compatibility. I have adopted your suggestion accordingly and added a ConsentConstraint, which provides the new protocol. The OnboardingDataSource now distinguishes the two constraints.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -166,17 +171,20 @@ public struct OnboardingConsentView: View {
         markdown: @escaping () async -> Data,
         action: @escaping () async -> Void,
         title: LocalizedStringResource? = LocalizationDefaults.consentFormTitle,
+        identifier: String = "DefaultConsentDocument",
```
Wondering here about the identifier. Maybe just name it 
```suggestion
        identifier: String = "ConsentDocument",
```

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -30,6 +33,7 @@ import SwiftUI
 ///         // The action that should be performed once the user has provided their consent.
 ///     },
 ///     title: "Consent",   // Configure the title of the consent view
+///     identifier: "MyFirstConsentForm", // Specify a unique identifier for the consent form, helpful for distinguishing consent forms when storing.
```
As noted, it would be great to use the enum approach in the docs to highlight this. Maybe we can even highlight this as a `tip` in the DocC documentation as a best practice?

RealLast replied:
I added a tip/explanation to the comments.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -90,7 +95,7 @@ public struct OnboardingConsentView: View {
                     if !willShowShareSheet {
                         Task { @MainActor in
                             /// Stores the finished PDF in the Spezi `Standard`.
-                            await onboardingDataSource.store(exportedConsentDocumented)
+                            await onboardingDataSource.store(exportedConsentDocumented, identifier: identifier)
                             await action()
                         }
```
I think it would make sense to do some error handling here, e.g., something like
```swift
Task {
    do {
        /// Stores the finished PDF in the Spezi `Standard`.
        try await onboardingDataSource.store(exportedConsentDocumented, identifier: identifier)
        await action()
    } catch {
        viewState = .base(.error(AnyLocalizedError(error: error)))
    }
}
```

In addition, we should make sure that the consent button is an async button that shows a progress indicator while the upload is happening and does indicate some processing state as the upload and PDF creation is working.

RealLast replied:
Thanks again for the suggestion! I have changed it accordingly. 
Also, I implemented the second suggestion regarding the async consent button, and I believe this one is actually very crucial.  I found that pressing the consent button multiple times rapidly sometimes causes odd behavior. I could skip most of the subsequent OnboardingViews by clicking the button multiple times.

To solve this, I have now added a new ViewState ".storing", which is active while the store function has not finished yet. During .storing, the "I consent" button is now deactivated, and a ProgressView overlay is shown. If the "store" function throws an error, the button is reactivated, allowing the user to try again. I also tested this manually by throwing an Error in ExampleStandard.store, but it could be worth writing a separate test case for this.

PSchmiedmayer reviewed (changes requested):
Thank you @RealLast for your first contribution to Spezi and tackling this issue. This has been a long-running limit of the onboarding module and is great to be addressed!

Sorry for the delay, we have been quite busy with a few end-of-June deadlines, sorry about this!

I have added a few comments across the PR with he majority focusing on how we can
1. Make this a backward-compatible change
2. Make sure that we use this as an opportunity to address some limitations that we currently have in the current implementation (e.g. error propagation from the standard to the view)

Apart from this, it would be great if you can address the failing UI tests on some platforms as we go though this PR. 

Please let us know if you have any follow-up questions based on the feedback and additional suggestions!

Thanks again for taking the time to make an open-source contribution!

RealLast commented:
Thank you very much @PSchmiedmayer for the detailed and thoughtful feedback! I went over it carefully and incorporated your suggestions. I will also add some considerations I had in mind to your comments.

Regarding the failing test, I saw that we got the following error on iPadOS. _"The request was denied by service delegate (SBMainWorkspace) for reason: NotFound ("Application "edu.stanford.spezi.onboarding.testappuitests.xctrunner" is unknown to FrontBoard")"_

I am not sure how to solve this, can I do anything on my side to fix this?


Thanks again!

Best,
Patrick

RealLast commented on Tests/UITests/TestApp/Views/OnboardingConsentMarkdownRenderingView1.swift:
Thanks for the input! I have changed it back to one view (class) which takes in the consentTitle, consentText, and documentIdentifier as parameters. I also added an "onboardingIdentifier" to the second consent view, so that both views can be distinguished by Spezi.

My only consideration is that I was unsure how to add a separate button for the second consent view in "OnboardingStartTestView.", even after having checked the underlying code in detail. Thus there is currently only one button, "Consent View (Markdown)," which shows the first consent document. The second document appears after the

PSchmiedmayer reviewed (changes requested):
Thank you for the improvements @RealLast!

I took a look at the iPad GitHub runner problem and it seems like the relevant iPad we used is no longer installed by default so I switched it to the newer iPad Air model to successfully start the test.

Still seems like some of the tests are failing on the CI which might be due to multiple changes in the UI and elements. The best way to identify what is going wrong is going to the checks of the PR (https://github.com/StanfordSpxezi/SpeziOnboarding/pull/51/checks) and downloading the `.xcresult` files that are provided as artifacts of the failed builds. They can be opened in Xcode and inspected (including a video recording of the failed test). These things can be frustrating so feel free to let me or @philippzagar know if you run into any issues or elements where you are not sure how to solve a failing test!

Apart from this: I took a look at the PR code and we are making some great progress! Thanks for all the improvements and incorporating this in the code base. I have looked at the current state and provided some additional feedback and suggestions based on the new additions and improvements. Most of them are now UI-state related and would be great to be incorporated with the latest improvements in the PR 🚀

Let me know if you have any questions or follow-up comments. I think after resolving those issues the PR should be ready to be merged, thanks for all the work 👍

RealLast commented on Tests/UITests/TestApp/Views/OnboardingConsentMarkdownRenderingView.swift:
```diff
@@ -13,10 +13,13 @@ import SwiftUI
 
 struct OnboardingConsentMarkdownRenderingView: View {
+    var consentTitle: String
+    var documentIdentifier: String
```
I have changed the properties to be constants (let), however they are still public, as they are passed during the construction of the views, e.g., [here](https://github.com/RealLast/SpeziOnboarding/blob/main/Tests/UITests/TestApp/OnboardingTestsView.swift#L28). I have taken inspiration from the [OnboardingTestViewNotIdentifiable](https://github.com/RealLast/SpeziOnboarding/blob/main/Tests/UITests/TestApp/Views/OnboardingTestViewNotIdentifiable.swift), which also uses a public property during initialization. I have changed the properties to private in the rest of the PR.

RealLast commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -85,13 +102,44 @@ public struct OnboardingConsentView: View {
+                            }
+                        }
+                    }
+                )
```
Thank you for the suggestions! I have incorporated and tested the changes accordingly. I also tried to produce some edge cases to see if anything breaks, and so far it seemed very stable in my tests. I would appreciate it if you could check again if the new implementation also looks good from your side!

RealLast commented:
Hi @PSchmiedmayer,

thanks again for the thoughtful feedback! I went over it carefully and have incorporated all of it, and hope I haven't overlooked anything. I also closed most of the conversations except two which I left open for potential further discussion. Let me know what you think :)

Thanks again!

Best

PSchmiedmayer commented on Sources/SpeziOnboarding/ConsentConstraint.swift:
```diff
@@ -0,0 +1,22 @@
+    /// - Parameters:
+    ///     - consent: The exported consent form represented as `PDFDocument` that should be added.            
+    ///     - identifier: A 'String' identifying the consent form as specified in `OnboardingConsentView`.                
+    func store(consent: PDFDocument, identifier: String) async throws
```
We have had some discussions in the office how to expand the overall setup of the consent document and one idea that came up was the question how we can avoid a breaking change in the future.
One idea that I think we should go for would be to package the consent document in a `ConsentDocumentExport` type that contains the PDF document (e.g. a `pdf` property), the identifier (the type can conform to `Identifiable`). This would allow is to pass this to this function and possibly add more elements to the type without introducing a breaking change as we would for a function signature that would get one more argument.

I would suggest that we pull this change in this transition to the `ConsentConstraint`.

I am also wondering if we should make the PDF property asynchronous (can be a simple wrapper that just exposes a private stored pdf property (e.g. `cachedPDF`) in an asynchronous computed property. While this might sounds strange at the beginning it would allow us to eventually lazily generate the PDF if it is needed by the application. Some apps might not even need the PDF and would just like to get the information from the Consent document and store it in a different way.

CC @AdritRao.

RealLast replied:
Thanks for the suggestion, I think that makes the whole implementation much more flexible! I have updated the code accordingly and introduced the `ConsentDocumentExport` type. I think now it's already much more future-proof :) I also moved the identifiers to a separate type `ConsentDocumentIdentifier` which conforms to `Identifiable`, and have updated the docs accordingly.

I also like the idea of making the pdf property async. I think that's a nice optimization for cases where users do not really need the PDF. I implemented it as you suggested. The `ConsentDocumentExport` can optionally receive a "cachedPDF" property during initialization, and `ConsentDocumentExport.pdf` would then return the cached value. In the `OnboardingConsentView`, this cachedPDF is always set by default when generating the ConsentDocumentExport.

However, in case a cachedPDF is not set, I currently return null for the document, as I have some considerations regarding the "lazy PDF export" as of now. To do that, we may need to store an instance of the `ConsentDocument` in the `ConsentDocumentExport`, to later call `ConsentDocument.export()`. I think this would possibly require making the `ConsentDocument` Sendable, as we might pass the `ConsentDocumentExport` between different actors. As the `ConsentDocument` is a View, I am not sure if that's the best approach. Thus, for now, I did not include this functionality and would like to hear your opinion first. We could also implement this in a future PR, and I could let the async "pdf" property throw something like an "UnimplementedError" for now.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -73,25 +92,42 @@ public struct OnboardingConsentView: View {
                 actionView: {
-                    OnboardingActionsView(
-                        LocalizedStringResource("CONSENT_ACTION", bundle: .atURL(from: .module)),
+                    AsyncButton(
```
Can probably just be a normal `Button` as we don't do any asynchronous actions here.

PSchmiedmayer commented on Sources/SpeziOnboarding/SpeziOnboarding.docc/ObtainingUserConsent.md:
````diff
@@ -28,10 +28,54 @@ OnboardingConsentView(
 )
 ```
 
+### Using multiple consent forms
````
Thank you for the added documentation!

PSchmiedmayer reviewed (changes requested):
Thank you @RealLast for the great changes and improvements; I think we are getting really close.

It seems like the UI tests are failing due to some changes to iOS 18, and I have pushed some elements for this in #52 that we should also incorporate here or merge the other PR first and then merge this one.

I have added some last comments with one larger one that handles future backward compatibility but should be easily doable.
We can keep the `OnboardingDataSource` as is for now, but we will then deprecate it in an other PR where we might introduce additional context to the consent document rendering and eventual export.

RealLast commented:
Thanks again @PSchmiedmayer for the suggestions! Much appreciated.
I have incorporated them and added a response to your last comment.
Also thanks for checking again the test runners! Curious to see whether they will run through now :)

PSchmiedmayer commented on Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport.swift:
```diff
@@ -0,0 +1,39 @@
+    /// - Parameters:
+    ///   - documentIdentfier: A unique `ConsentDocumentIdentifier` identifying the exported `ConsentDocument`.
+    ///   - cachedPDF: A `PDFDocument` representing exported from a `ConsentDocument`. Optional parameter which will be stored internally and is accessible via the async property `ConsentDocumentExport.pdf`.
+    public init(
```
I would keep this initializers internal for now, as we might want to change it it the future when we enable an async loading of the pdf.
```suggestion
   init(
```

PSchmiedmayer commented on Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport.swift:
```diff
@@ -0,0 +1,39 @@
+    ///   - cachedPDF: A `PDFDocument` representing exported from a `ConsentDocument`. Optional parameter which will be stored internally and is accessible via the async property `ConsentDocumentExport.pdf`.
+    public init(
+        documentIdentifier: ConsentDocumentIdentifier,
+        cachedPDF: PDFDocument? = nil
```
I would make this non-optional for now, would also avoid making the `pdf` property optional and therefore simplify it for now. This might become more sophisticated in the future when we introduce a lazy PDF rendering.

PSchmiedmayer commented on Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport.swift:
```diff
@@ -0,0 +1,39 @@
+import PDFKit
+
+/// A type representing an exported `ConsentDocument`. It holds the exported `PDFDocument` and the corresponding `ConsentDocumentIdentifier`.
+public actor ConsentDocumentExport {
```
Thank you for introducing this type here; looks nice!

As noted on the identifier type, we don't really need the String indirection for now so I would suggest to keep it simply a string for now.

PSchmiedmayer commented on Sources/SpeziOnboarding/ConsentView/ConsentDocumentIdentifier.swift:
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+/// A type representing an unique identifier for a `ConsentDocument`. Allows to distinguish multiple documents during `ConsentConstraint.store`.
+public struct ConsentDocumentIdentifier: Identifiable, Sendable, Equatable {
```
Would be great to learn your reasoning about the wrapper type here. As far as I can see working directly with a string might make things a bit easier (also easier for a Codable conformance and more) but happy to be convinced otherwise 👍

RealLast replied:
I agree, it is not really needed. I have reverted it back to Strings :)
My initial thought was to make it more strictly typed and adhere to the Identifiable protocol, but I agree that simple Strings make it easier.

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConsentView.swift:
```diff
@@ -116,10 +116,14 @@ public struct OnboardingConsentView: View {
                                 try await onboardingDataSource.store(
-                                    exportedConsentDocumented,
-                                    identifier: identifier
+                                   documentExport
```
```suggestion
                                    documentExport
```

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingDataSource.swift:
```diff
@@ -50,11 +50,16 @@ public class OnboardingDataSource: Module, EnvironmentAccessible {
     /// - Parameter consent: The exported consent form represented as `PDFDocument` that should be added.
     /// - Parameter identifier: A 'String' identifying the consent form as specified in OnboardingConsentView.                
-    public func store(_ consent: PDFDocument, identifier: String) async throws {
+    public func store(_ consent: ConsentDocumentExport) async throws {
```
This would technically be a breaking change of the public interface here; I would therefore rather suggest to keep the function definition the old way and rather map it to the new `ConsentDocumentExport` here when needed. Also avoids us doing one async call here.

RealLast replied:
That makes sense, so I changed it to the previous version.
But, just to make sure: This PR already is changing the public interface anyways, before my changes, it was:
``` swift
public func store(_ consent: PDFDocument) async throws {
```

And in previous commits to this PR, I already introduced the **identifier** as new parameter:
``` swift
public func store(_ consent: PDFDocument, identifier: String) async throws {
```

Thus, the interface would currently change anyway, right? To avoid this, we could possibly make the identifier optional and set it to the same default value as the identifier in the [OnboardingConsentView](https://github.com/RealLast/SpeziOnboarding/blob/main/Sources/SpeziOnboarding/OnboardingConsentView.swift#L227). Not sure if this is relevant, just wanted to make sure that we do not overlook something :)

PSchmiedmayer replied:
Very good point, I missed this! Agree, the identifier should have a default value, if we have that in place we don't even need to make it optional.

Great catch!

PSchmiedmayer reviewed (changes requested):
@RealLast Thank you for the further improvements; the major last element would be a removal of a breaking change + suggestions to structure the ConsentDocumentExport types.

After that I see this PR as ready to be merged; thanks for all the time and effort that went into the PR!

RealLast commented:
Thanks @PSchmiedmayer, I have incorporated the suggestions! :)

RealLast commented on Sources/SpeziOnboarding/OnboardingDataSource.swift:
```diff
@@ -34,18 +34,29 @@ import SwiftUI
-        Task { @MainActor in
-            await standard.store(consent: consent)
+    /// - Parameter consent: The exported consent form represented as `ConsentDocumentExport` that should be added.
+    public func store(_ consent: PDFDocument, identifier: String) async throws {
```
I added the default value to the identifier. I wanted to make sure, however, that the default identifier is the same one as the one used in OnboardingConsentView, so I added a new public constant:
``` swift
public struct OnboardingConsentView : View {
    ...
    public static let defaultDocumentIdentifier = "ConsentDocument"
}
```

Let me know if we should use anything else than a global static constant for the default identifier.

PSchmiedmayer replied:
Good idea to ensure that we use the same identifier across the board. In general we use the the following mechanism for elements like this:
```swift
public actor ConsentDocumentExport {
    public enum Defaults {
        public static let documentIdentifier = "ConsentDocument"
    }
    
    // ...
```

This allows you to easily get it by calling `ConsentDocumentExport.Defaults.documentIdentifier`. By also packaging it there it is in the same place as where the identifier will normally be stored and not scoped to a View bur rather on the model level.

RealLast replied:
Great, thanks a lot! I have included this into the `ConsentDocumentExport`.

PSchmiedmayer replied:
Thank you! 🚀

RealLast commented:
Thank you very much @PSchmiedmayer, excited to see the code getting merged into Spezi!
I have added the default identifier.

Regarding the failing tests, to me it looks like a performance issue, or in general an issue with the runners. I attached a screenshot of the error from the xcresult file:

![test_failure](https://github.com/user-attachments/assets/d9368230-2c37-4403-bc56-a502eb2ab642)

PSchmiedmayer commented:
Thanks for taking a look at the tests; I will try to investigate this later this week!

RealLast commented:
Thanks! I also have incorporated the last changes regarding the default document identifier.

PSchmiedmayer commented:
@RealLast We have worked on some UI testing infrastructure improvements in #55 and once this is merged we can pull this in to this PR and merge it from there 👍

PSchmiedmayer commented:
@RealLast We have merged the fixed for the builds on the main branch and have also updated out build agents; it would be amazing if you can merge main to this branch to eventually merge this PR; let me know if you have any questions or run into any issues!

PSchmiedmayer commented on Sources/SpeziOnboarding/OnboardingConstraint.swift:
```diff
@@ -23,15 +23,7 @@ import Spezi
-        """
-    )
+    /// - Parameter consent: The exported consent form represented as `PDFDocument` that should be added.
+    @MainActor
```
I think we have lost some of the deprecation warnings and documentation in the merge:
```suggestion
    /// - Parameters:
    ///     - consent: The exported consent form represented as `PDFDocument` that should be added.
    @available(
        *,
        deprecated,
        message: """
        Storing consent documents without an identifier is deprecated.
        Please use `ConsentConstraint.store(consent: PDFDocument, identifier: String)` instead.
        """
    )
    @MainActor
```

RealLast replied:
Whoops, sorry about that!

PSchmiedmayer replied:
No worries; happens in merge commits all the time 👍

PSchmiedmayer commented on Tests/UITests/TestAppUITests/SpeziOnboardingTests.swift:
You can remove the `file_length` disable of SwiftLint in this file.

PSchmiedmayer reviewed (approved):
@RealLast Looks great, all the builds are passing now!

I only had tow very small comment that occurred in the merge, once these are resolved we can merge the PR. The CodeCov element can't pass as codecov changed their logic that breaks public forks but we can merge it despite that.

Feel free to tag me once you have resolved the last comments and I can then merge the PR with my admin permissions.

Thanks for all the work here that went into the PRs and improving Spezi Onboarding!

RealLast commented:
Thanks @PSchmiedmayer for checking it, I have resolved the comment! Waiting for the merge :)
